libcontainer/cgroups/utils.go

```diff
@@ -415,23 +415,23 @@ func ConvertCPUSharesToCgroupV2Value(cpuShares uint64) uint64 {
 
 // ConvertMemorySwapToCgroupV2Value converts MemorySwap value from OCI spec
 // for use by cgroup v2 drivers. A conversion is needed since Resources.MemorySwap
-// is defined as memory+swap combined, while in cgroup v2 swap is a separate value.
+// is defined as memory+swap combined, while in cgroup v2 swap is a separate value,
+// so we need to subtract memory from it where it makes sense.
 func ConvertMemorySwapToCgroupV2Value(memorySwap, memory int64) (int64, error) {
-	// for compatibility with cgroup1 controller, set swap to unlimited in
+	// For compatibility with cgroup1 controller, set swap to unlimited in
 	// case the memory is set to unlimited, and swap is not explicitly set,
 	// treating the request as "set both memory and swap to unlimited".
 	if memory == -1 && memorySwap == 0 {
 		return -1, nil
```

Comment:
So what this basically means is that I don't have a way to set the swap limit to 0 if the memory limit is set to unlimited.

This seems a bit confusing. If I'm trying to set the swap limit to 0 while allowing unlimited memory for an important container, I end up with a container that has unlimited swap instead.

Reply:
First, this is not really related to this PR.

Second, because OCI has two fields -- memory and memory+swap (https://github.com/opencontainers/runtime-spec/blob/8f3fbc881602d85699e5c448634ec1288860d966/specs-go/config.go#L378) -- there is no way to set unlimited memory and zero swap. If you want no swap, either disable it on the host, or set containers' memory and memory+swap limit to the same (limited) value. To me, doing either is a bad idea.

Third, not allowing container memory to be swapped out may result in OOM killings which is way worse than swapping out. Having said that, setting the memory to unlimited value may result in the same thing, only later (if swap is set).

To reiterate -- swap is good for you, as the only alternative is OOM kill.

Reply:
Update: if you still want it, you can set memory to unlimited and swap to zero on cgroup v2 by using unified resources (https://github.com/opencontainers/runtime-spec/blob/main/config-linux.md#unified), i.e. set `unified.memory.swap.max` to `0`.

Reply:
Something like this

```json
{ 
  "unified": {
      "memory.max": "-1",
      "memory.swap.max": "0"
  }
}
```

can be fed to `runc update -r` (if you want to modify resources during runtime).

For `runc create/start`, a similar addition to `config.json` is needed.